src/fprime/fbuild/interaction.py

```diff
@@ -69,41 +76,98 @@ def run_impl(deployment: Path, path: Path, platform: str, verbose: bool):
-    """Adds new component to CMakeLists.txt"""
+    """ Adds new component or port to CMakeLists.txt"""
+    print("[INFO] Found CMakeLists.txt at '{}'".format(list_file))
+    with open(list_file, "r") as f:
+        lines = f.readlines()
+        index = 0
+        while "add_fprime_subdirectory" not in lines[index]:
+            index += 1
+        while "add_fprime_subdirectory" in lines[index]:
+            index += 1
+
+    if not confirm(
+        "Add component {} to {} {}?".format(
+            comp_path, list_file, "at end of file before topology inclusion?"
+        )
+    ):
+        return False
+
+    addition = (
+        'add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/' + str(comp_path) + '/")\n'
+    )
+    if addition in lines:
+        print("Component already added to CMakeLists.txt")
+        return True
+    lines.insert(index, addition)
+    with open(list_file, "w") as f:
+        f.write("".join(lines))
+    return True
+
+
+def regenerate(cmake_list_file):
+    # Ensures build directory exists, then refreshes build cache
+    build_dir = Path(cmake_list_file.parent, "build-fprime-automatic-native")
+    if not os.path.isdir(build_dir):
+        raise CMakeInvalidBuildException(build_dir)
+    handler = CMakeHandler()
+    print("Refreshing cache to include new addition")
+    handler._cmake_refresh_cache(build_dir)
+
+
+def add_unit_tests(deployment, comp_path, platform, verbose):
+    # Creates unit tests and moves them into test/ut directory
+    os.chdir(str(comp_path))
+    if confirm("Would you like to generate unit tests?: "):
+        target = Target.get_target("impl", {"ut"})
+        build = Build(target.build_type, deployment, verbose=verbose)
+        build.load(comp_path, platform)
+        build.execute(target, context=comp_path, make_args={})
+        test_files = [
+            "Tester.hpp",
+            "Tester.cpp",
+            "TesterBase.hpp",
+            "TesterBase.cpp",
+            "GTestBase.hpp",
+            "GTestBase.cpp",
+            "TestMain.cpp",
+        ]
+        for file in test_files:
+            if os.path.isfile(file):
+                new_name = Path("test", "ut", file)
+                os.rename(file, str(new_name))
+    else:
+        shutil.rmtree("test")
+
+
+def add_port_to_cmake(list_file: Path, comp_path: Path):
+    """ Adds new port to CMakeLists.txt in port directory"""
     print("[INFO] Found CMakeLists.txt at '{}'".format(list_file))
     with open(list_file, "r") as file_handle:
         lines = file_handle.readlines()
-    topology_lines = [
-        (line, text) for line, text in enumerate(lines) if "/Top/" in text
-    ]
-    line = len(topology_lines)
-    if topology_lines:
-        line, text = topology_lines[0]
-        print(
-            "[INFO] Topology inclusion '{}' found on line {}.".format(
-                text.strip(), line + 1
-            )
-        )
+    index = 0
+    line = lines[index]
+    while line != "set(SOURCE_FILES\n":
```

Comment:
We might want to make this a regular expression comparison rather than a direct string comparison.

Otherwise we'll fail to match in the case of something like `set( SOURCE_FILES` where a space is added.